RenderStateNotationParser/src/RenderStateNotationParserImpl.cpp

```diff
@@ -26,131 +26,154 @@
 }
 
-static void Deserialize(const nlohmann::json& Json, RTGeneralShaderGroupNotation& Type, DynamicLinearAllocator& Allocator)
+static void Deserialize(const nlohmann::json& Json, RTGeneralShaderGroupNotation& Type, DynamicLinearAllocator& Allocator, const InlineStructureCallbacks& Callbacks)
 {
-    if (Json.contains("Name"))
-        Deserialize(Json["Name"], Type.Name, Allocator);
-
-    if (Json.contains("pShader"))
-        Deserialize(Json["pShader"], Type.pShaderName, Allocator);
+    Deserialize(Json.at("Name"), Type.Name, Allocator);
+    Callbacks.ShaderCallback(Json.at("pShader"), Type.pShaderName, Allocator);
 }
 
-static void Deserialize(const nlohmann::json& Json, RTTriangleHitShaderGroupNotation& Type, DynamicLinearAllocator& Allocator)
+static void Deserialize(const nlohmann::json& Json, RTTriangleHitShaderGroupNotation& Type, DynamicLinearAllocator& Allocator, const InlineStructureCallbacks& Callbacks)
 {
-    if (Json.contains("Name"))
-        Deserialize(Json["Name"], Type.Name, Allocator);
+    Deserialize(Json.at("Name"), Type.Name, Allocator);
 
     if (Json.contains("pClosestHitShader"))
-        Deserialize(Json["pClosestHitShader"], Type.pClosestHitShaderName, Allocator);
+        Callbacks.ShaderCallback(Json["pClosestHitShader"], Type.pClosestHitShaderName, Allocator);
 
     if (Json.contains("pAnyHitShader"))
-        Deserialize(Json["pAnyHitShader"], Type.pAnyHitShaderName, Allocator);
+        Callbacks.ShaderCallback(Json["pAnyHitShader"], Type.pAnyHitShaderName, Allocator);
 }
 
-static void Deserialize(const nlohmann::json& Json, RTProceduralHitShaderGroupNotation& Type, DynamicLinearAllocator& Allocator)
+static void Deserialize(const nlohmann::json& Json, RTProceduralHitShaderGroupNotation& Type, DynamicLinearAllocator& Allocator, const InlineStructureCallbacks& Callbacks)
 {
-    if (Json.contains("Name"))
-        Deserialize(Json["Name"], Type.Name, Allocator);
+    Deserialize(Json.at("Name"), Type.Name, Allocator);
 
     if (Json.contains("pIntersectionShader"))
-        Deserialize(Json["pIntersectionShader"], Type.pIntersectionShaderName, Allocator);
+        Callbacks.ShaderCallback(Json["pIntersectionShader"], Type.pIntersectionShaderName, Allocator);
 
     if (Json.contains("pClosestHitShader"))
-        Deserialize(Json["pClosestHitShader"], Type.pClosestHitShaderName, Allocator);
+        Callbacks.ShaderCallback(Json["pClosestHitShader"], Type.pClosestHitShaderName, Allocator);
 
     if (Json.contains("pAnyHitShader"))
-        Deserialize(Json["pAnyHitShader"], Type.pAnyHitShaderName, Allocator);
+        Callbacks.ShaderCallback(Json["pAnyHitShader"], Type.pAnyHitShaderName, Allocator);
 }
 
-static void Deserialize(const nlohmann::json& Json, RayTracingPipelineNotation& Type, DynamicLinearAllocator& Allocator)
+static void Deserialize(const nlohmann::json& Json, RayTracingPipelineNotation& Type, DynamicLinearAllocator& Allocator, const InlineStructureCallbacks& Callbacks)
```

Comment:
All versions of the Deserialize method take Json, Allocator and Callbacks arguments.
It makes sense to make these methods member of a class or a struct that contains these members so   that you don't have to pass the same arguments every time.

One way is to make a functor `Deserializer` with overloaded versions of `operator()`:
```cpp
struct (SomeKindOf)Parser // Deserialize is not quite appropriate here
{
  const nlohmann::json& Json;
  DynamicLinearAllocator& Allocator; // Maybe allocator can be a member, not a reference
  std::function<void(const nlohmann::json&, const char*&, DynamicLinearAllocator&)> ShaderCallback;
  std::function<void(const nlohmann::json&, const char*&, DynamicLinearAllocator&)> RenderPassCallback;
  std::function<void(const nlohmann::json&, const char*&, DynamicLinearAllocator&)> ResourceSignatureCallback;

  void operator()(PipelineStateNotation& Type);
  void operator()(GraphicsPipelineNotation& Type);
  //...
};

Reply:
We have unique `const nlohmann::json&` for each call. 
I think it's best to leave it as it is for now.